spark-operator/src/main/java/org/apache/spark/kubernetes/operator/reconciler/SparkApplicationReconciler.java

```diff
@@ -0,0 +1,227 @@
+import org.apache.spark.kubernetes.operator.controller.SparkApplicationContext;
+import org.apache.spark.kubernetes.operator.health.SentinelManager;
+import org.apache.spark.kubernetes.operator.reconciler.observers.AppDriverReadyObserver;
+import org.apache.spark.kubernetes.operator.reconciler.observers.AppDriverRunningObserver;
+import org.apache.spark.kubernetes.operator.reconciler.observers.AppDriverStartObserver;
+import org.apache.spark.kubernetes.operator.reconciler.observers.AppDriverTimeoutObserver;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.AppCleanUpStep;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.AppInitStep;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.AppReconcileStep;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.AppResourceObserveStep;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.AppRunningStep;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.AppTerminatedStep;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.AppValidateStep;
+import org.apache.spark.kubernetes.operator.reconciler.reconcilesteps.UnknownStateStep;
+import org.apache.spark.kubernetes.operator.utils.ApplicationStatusUtils;
+import org.apache.spark.kubernetes.operator.utils.LoggingUtils;
+import org.apache.spark.kubernetes.operator.utils.StatusRecorder;
+
+import static org.apache.spark.kubernetes.operator.reconciler.ReconcileProgress.completeAndDefaultRequeue;
+import static org.apache.spark.kubernetes.operator.reconciler.SparkReconcilerUtils.commonResourceLabelsStr;
+
+/**
+ * Reconciler for Spark Application.
+ * Performs sanity check on the app, identify the reconcile steps based on App status
+ * and execute the steps.
+ */
+@ControllerConfiguration
+@Slf4j
+@RequiredArgsConstructor
+public class SparkApplicationReconciler
+    implements Reconciler<SparkApplication>,
+    ErrorStatusHandler<SparkApplication>,
+    EventSourceInitializer<SparkApplication>,
+    Cleaner<SparkApplication> {
+  private final StatusRecorder statusRecorder;
+  private final SentinelManager<SparkApplication> sentinelManager;
+
+  @Override
+  public UpdateControl<SparkApplication> reconcile(SparkApplication sparkApplication,
+                                                   Context<SparkApplication> context)
+      throws Exception {
+    LoggingUtils.TrackedMDC trackedMDC = new LoggingUtils.TrackedMDC();
+    try {
+      trackedMDC.set(sparkApplication);
+      if (sentinelManager.handleSentinelResourceReconciliation(sparkApplication,
+          context.getClient())) {
+        return UpdateControl.noUpdate();
+      }
+      log.debug("Start reconciliation.");
+      statusRecorder.updateStatusFromCache(sparkApplication);
+      SparkApplicationContext ctx = new SparkApplicationContext(sparkApplication, context);
+      List<AppReconcileStep> reconcileSteps = getReconcileSteps(sparkApplication);
```

Comment:
This is very interesting with the step abstraction. Just one note one this, I did not dive too deep into this but, JOSDK has an abstraction descript resources reconciliations and and workflows on top of that, might be interesting to see if these steps could be fitted in that model, that might spare you some code. Also it mainly provides lot's of optimizations regarding reconciliation cycles, see:
https://javaoperatorsdk.io/docs/dependent-resources
https://javaoperatorsdk.io/docs/workflows